main/src/main/java/cgeo/geocaching/utils/OfflineTranslateUtils.java

```diff
@@ -127,21 +132,38 @@ public static void getTranslator(final Activity activity, final Language sourceL
     }
 
     public static void translateParagraph(final Translator translator, final OfflineTranslateUtils.Status status, final String text, final Consumer<String> consumer, final Consumer<Exception> errorConsumer) {
-        final List<String> origText = Arrays.asList(text.split("\n"));
-        final List<String> translatedText = new ArrayList<>();
-        for (String p : origText) {
-            translator.translate(p)
-                    .addOnSuccessListener(translation -> {
-                        translatedText.add(translation);
-                        if (origText.size() == translatedText.size()) {
-                            consumer.accept(String.join("\n", translatedText));
-                            status.updateProgress();
-                        }
-                    })
-                    .addOnFailureListener(e -> {
-                        status.abortTranslation();
-                        errorConsumer.accept(e);
-                    });
+        final Document document = Jsoup.parse(text);
+        final Elements elements = document.children().select("*");
+        final AtomicInteger remaining = new AtomicInteger(elements.size());
+        for (Element element : elements) {
+            final String current = element.ownText();
```

Comment:
You extract text using `ownText()`...

Reply:
`text()` would include the text of all siblings, which to my understanding would break formatting.

Reply:
Using text would indeed break formatting.

But using ownText could also lead to unwanted results. Eg for the example in JSoup documentation:

![Screenshot_2025-05-04-09-56-10-30_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://github.com/user-attachments/assets/3597b658-dcc4-495a-930d-b471506fe093)

Using `ownText` and `text` as you do right now would lead to `<p>Hello now!<p>` (removing the bold `there`), wouldn't it? (Can't try, not on my machine)

Reply:
Come to think of it, keeping formatting in translation is far from trivial anyway. Because order of words and their meaning can of course change during translation. Eg if sentence "Hello there now!" would be translated to "Hallo jetzt da draußen!", then how should we know that now the last two words must be bold instead of the middle one?

Reply:
Switching to translating just text nodes instead of elements resolved the `text()` / `ownText()` issue, but comes at the cost of having to translate incomplete sentences sometimes (which I had though my original approach already would do). Let's see how it works in real life.

Reply:
> Switching to translating just text nodes instead of elements resolved the `text()` / `ownText()` issue, but comes at the cost of having to translate incomplete sentences sometimes (which I had though my original approach already would do). Let's see how it works in real life.

Tbh I would personally prefer translating whole sentences (and loosing formatting) over keeping format (and translating sentence parts only). This is up to debate of course.

What about neglecting basic formats like bold/italic/underline in favor of whole sentences, but keeping separate trandlations for list elements and in table cells?